apps/full-site-editing/full-site-editing-plugin/starter-page-templates/class-wp-rest-batch-controller.php

```diff
@@ -0,0 +1,160 @@
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * This handles the building of the response for the batch requests we make.
+	 *
+	 * @param WP_REST_Request $request The current matched request object.
+	 * @return WP_REST_Response A collection of response data for batch endpoints.
+	 */
+	public function handle_batch_requests( $request ) {
+		$data = [];
+
+		// Foreach request specified in the requests param, run the endpoint.
+		foreach ( $request['requests'] as $request_params ) {
+			$response     = $this->handle_request( $request_params );
+			$key          = $request_params['method'] . ' ' . $request_params['route'];
+			$data[ $key ] = $this->prepare_for_collection( $response );
+		}
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * This handles the building of the response for the batch requests we make.
+	 *
+	 * @param array $request_params Data to build a WP_REST_Request object from.
+	 * @return WP_REST_Response Response data for the request.
+	 */
+	public function handle_request( $request_params ) {
+		$request = new WP_REST_Request( $request_params['method'], $request_params['route'] );
+
+		// Add specified request parameters into the request.
+		if ( ! empty( $request_params['params'] ) ) {
+			foreach ( $request_params['params'] as $param_name => $param_value ) {
+				$request->set_param( $param_name, $param_value );
+			}
+		}
+
+		return rest_do_request( $request );
+	}
+
+	/**
+	 * Prepare a response for inserting into a collection of responses.
+	 *
+	 * @param WP_REST_Response $response Response object.
+	 * @return array|WP_REST_Response Response data, ready for insertion into collection data.
+	 */
+	public function prepare_for_collection( $response ) {
+		if ( ! ( $response instanceof WP_REST_Response ) ) {
+			return $response;
+		}
+
+		$data   = (array) $response->get_data();
+		$server = rest_get_server();
+
+		if ( method_exists( $server, 'get_compact_response_links' ) ) {
+			$links = call_user_func( [ $server, 'get_compact_response_links' ], $response );
+		} else {
+			$links = call_user_func( [ $server, 'get_response_links' ], $response );
+		}
+
+		if ( ! empty( $links ) ) {
+			$data['_links'] = $links;
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Validates requests to be processed.
+	 *
+	 * @param array $requests Array of requests to process.
+	 * @return bool|\WP_Error
+	 */
+	public function validate_requests( $requests ) {
+		// If requests isn't an array of requests then we don't process the batch.
+		if ( ! is_array( $requests ) ) {
+			return new WP_Error( 'rest_invalid_param', 'The requests parameter must be an array of requests.', [ 'status' => 400 ] );
+		}
+
+		foreach ( $requests as $request ) {
+			// If the method or route is not set then we do not run the requests.
+			if ( ! isset( $request['method'] ) || ! isset( $request['route'] ) ) {
+				return new WP_Error( 'rest_invalid_param', 'You must specify the method and route for each request.', [ 'status' => 400 ] );
+			}
+
```

Comment:
we need to check if `params` is defined before to check its form. If not it will trigger a PHP Notice:

```
[25-Jul-2019 12:55:33 UTC] PHP Notice:  Undefined index: type in /var/www/html/wp-includes/rest-api.php on line 1141
[25-Jul-2019 12:55:33 UTC] PHP Stack trace:
[25-Jul-2019 12:55:33 UTC] PHP   1. {main}() /var/www/html/index.php:0
[25-Jul-2019 12:55:33 UTC] PHP   2. require() /var/www/html/index.php:17
[25-Jul-2019 12:55:33 UTC] PHP   3. wp() /var/www/html/wp-blog-header.php:16
[25-Jul-2019 12:55:33 UTC] PHP   4. WP->main() /var/www/html/wp-includes/functions.php:1105
[25-Jul-2019 12:55:33 UTC] PHP   5. WP->parse_request() /var/www/html/wp-includes/class-wp.php:737
[25-Jul-2019 12:55:33 UTC] PHP   6. do_action_ref_array() /var/www/html/wp-includes/class-wp.php:387
[25-Jul-2019 12:55:33 UTC] PHP   7. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:531
[25-Jul-2019 12:55:33 UTC] PHP   8. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[25-Jul-2019 12:55:33 UTC] PHP   9. rest_api_loaded() /var/www/html/wp-includes/class-wp-hook.php:286
[25-Jul-2019 12:55:33 UTC] PHP  10. WP_REST_Server->serve_request() /var/www/html/wp-includes/rest-api.php:309
[25-Jul-2019 12:55:33 UTC] PHP  11. WP_REST_Server->dispatch() /var/www/html/wp-includes/rest-api/class-wp-rest-server.php:329
[25-Jul-2019 12:55:33 UTC] PHP  12. WP_REST_Request->sanitize_params() /var/www/html/wp-includes/rest-api/class-wp-rest-server.php:888
[25-Jul-2019 12:55:33 UTC] PHP  13. rest_parse_request_arg() /var/www/html/wp-includes/rest-api/class-wp-rest-request.php:793
[25-Jul-2019 12:55:33 UTC] PHP  14. rest_validate_request_arg() /var/www/html/wp-includes/rest-api.php:1002
[25-Jul-2019 12:55:33 UTC] PHP  15. rest_validate_value_from_schema() /var/www/html/wp-includes/rest-api.php:965
[25-Jul-2019 12:55:33 UTC] PHP  16. rest_validate_value_from_schema() /var/www/html/wp-includes/rest-api.php:1150
```

Something like this?

```php
	if ( ! isset( $request['params'] ) ) {
		return new WP_Error( 'rest_invalid_param', 'You must specify the params for each request.', [ 'status' => 400 ] );
	}
```

Reply:
Could you paste your request path here that triggered that?

The notice seems to point to a missing `type` parameter in the endpoint schema, but all args that I can see do have that.

Reply:
Yup, you're right. I thought that `params` was a required attribute but it isn't. However, wondering id we should to handle the `PHP Notice:  Undefined index: type` notice.